fontforge/svg.c

```diff
@@ -766,60 +766,75 @@ SplineChar *SCHasSubs(SplineChar *sc, uint32 tag) {
 	if ( pst->type==pst_substitution &&
 		FeatureTagInFeatureScriptList(tag,pst->subtable->lookup->features) )
 return( SFGetChar(sc->parent,-1,pst->u.subs.variant));
+
     }
 return( NULL );
 }
 
+char *GetLigature(SplineChar *sc) {
+    PST *best = HasLigature(sc);
+    if ( best != NULL ) {
+    	int32 univals[50];
+        char *buf = "";
+        int i, c;
+		c = LigCnt(sc->parent,best,univals,sizeof(univals)/sizeof(univals[0]));
+		for ( i=0; i<c; ++i ) {
+		    if ( (univals[i]>='A' && univals[i]<='Z') ||  (univals[i]>='a' && univals[i]<='z') || (univals[i] >= '0' && univals[i] <= '9') || univals[i] == '_' || univals[i] == '-')
+		    	asprintf(&buf,"%s%c", buf, univals[i]);
+		    else
+		    	asprintf(&buf, "%s&#x%x;", buf, (unsigned int) univals[i]);
+		}
+        return (  copy(buf)  );
```

Comment:
Seems like this won't be freed based on how the function is called.

Reply:
Some of our formatting functions use a static fixed-length array to get around this problem. One returns a pointer to the static value, which is retained until it gets written over by the next call. Because there's just the one buffer there's no need to free it (indeed, there's a need *not* to free it). 

It's not the greatest style but it might be the easiest solution for this case.

Reply:
I'm not a fan of that approach (of using a static fixed length buffer), but a lot of this code could be simplified by just using smprintf (and remembering to free the result afterwards)

Reply:
I also prefer dynamic allocation where practical. There are a lot of cases in which it is impossible to allocate a static buffer that is long enough for the top 0.1% but not ridiculously wasteful the rest of the time.

Reply:
I would really appreciate if you could help to re-write this code in right way as c lang is not my main syntax.

Reply:
refactored